### PR TITLE
Atomize RenameFile in KeyManagedEncryptedEnv

### PR DESCRIPTION
**Pull request by hunterlxt**
fix bug: https://github.com/tikv/tikv/issues/9115

Summary: we need to update encryption metadata via `encryption::DataKeyManager`, which cannot combine with the actual file operation into one atomic operation. In `RenameFile`, when the src_file has been removed, power is off, then we may lost the file info of src_file next restart.

Signed-off-by: Xintao <hunterlxt@live.com>

### CODE REVIEW

**Comment by hunterlxt** [comment]
@tabokie @Connor1996  PTAL

**Review by tabokie — APPROVED** [review]
Looks good, but I want to make sure future implementation won't embed critical logic inside `Rename` method, e.g. file-specific state transfer. Maybe add a comment or simply deprecate the function in `KeyManager`?

**Review by hicqu — APPROVED** [review]

File: encryption/encryption.cc [hunk 1]
```diff
@@ -434,14 +434,18 @@ Status KeyManagedEncryptedEnv::LinkFile(const std::string& src_fname,
...
+  // Link(copy)File instead of RenameFile to avoid losing src_fname info when
+  // failed to rename the src_fname in the file system.
+  Status s = key_manager_->LinkFile(src_fname, dst_fname);
   if (!s.ok()) {
     return s;
   }
   s = target()->RenameFile(src_fname, dst_fname);
-  if (!s.ok()) {
+  if (s.ok()) {
+    key_manager_->DeleteFile(src_fname);
```
Comment (yiwu-arbug) [inline encryption/encryption.cc:445]:
```suggestion
    s = key_manager_->DeleteFile(src_fname);
```

**Review by yiwu-arbug — CHANGES_REQUESTED** [review]

**Comment by yiwu-arbug** [comment]
> Looks good, but I want to make sure future implementation won't embed critical logic inside `Rename` method, e.g. file-specific state transfer. Maybe add a comment or simply deprecate the function in `KeyManager`?

Great suggestion. Let's remove `KeyManager::RenameFile` in encryption.h.

**Comment by yiwu-arbug** [comment]
We need similar changes for TiKV logic as well.

**Comment by yiwu-arbug** [comment]
revert redundant formatter change?

File: encryption/encryption.cc [hunk 2]
```diff
@@ -432,7 +432,8 @@ Status KeyManagedEncryptedEnv::LinkFile(const std::string& src_fname,
   s = target()->LinkFile(src_fname, dst_fname);
   if (!s.ok()) {
     // Ignore error
-    key_manager_->DeleteFile(dst_fname);
+    Status delete_status __attribute__((__unused__)) =
+        key_manager_->DeleteFile(dst_name);
```
Comment (yiwu-arbug) [inline encryption/encryption.cc:436]: `assert(delete_status.ok())`

File: encryption/encryption.cc [hunk 3]
```diff
@@ -426,22 +431,28 @@ Status KeyManagedEncryptedEnv::LinkFile(const std::string& src_fname,
   }
   s = target()->LinkFile(src_fname, dst_fname);
   if (!s.ok()) {
-    // Ignore error
-    key_manager_->DeleteFile(dst_fname);
+    Status delete_status __attribute__((__unused__)) =
+        key_manager_->DeleteFile(dst_fname);
+    assert(delete_status.ok())
```
Comment (yiwu-arbug) [inline encryption/encryption.cc:436]:
```suggestion
    assert(delete_status.ok());
```

File: encryption/encryption.cc [hunk 4]
```diff
@@ -426,22 +431,28 @@ Status KeyManagedEncryptedEnv::LinkFile(const std::string& src_fname,
...
   s = target()->RenameFile(src_fname, dst_fname);
-  if (!s.ok()) {
-    // Ignore error
-    key_manager_->RenameFile(dst_fname, src_fname);
+  if (s.ok()) {
+    s = key_manager_->DeleteFile(src_fname);
+  } else {
+    Status delete_status __attribute__((__unused__)) =
+        key_manager_->DeleteFile(dst_fname);
+    assert(delete_status.ok())
```
Comment (yiwu-arbug) [inline encryption/encryption.cc:455]:
```suggestion
    assert(delete_status.ok());
```

**Review by yiwu-arbug — APPROVED** [review]
LGTM

**Pull request merged**